### DISCUSSION TITLE
Relative markdown link following for better GitHub integration?

### DISCUSSION THREAD

**Discussion by Woodshape** [epwalsh/obsidian.nvim#287] [Q&A] [answered]
Hello everybody,

I have the following problem. I want to set up my obsidian.nvim to integrate nicely with GitHub as a frontend.

Assuming this project structure

- root (.obsidian's home)
  - projects
    - some-link.md
    - some-other-link.md

Inside _some-link.md_ I have a forward link to _some-other-link.md_ using the markdown syntax [some-other-link](some-other-link.md) **(some-other-link.md)**. This is required to be relative to the path of _some-link.md_ because of how GitHub handles links internally. So **(projects/some-other-link.md)** will not work with GitHub and leads to GitHub linking **.../blob/projects/projects/some-other-link.md** to produce a 404.

As far as I understand, using the relative link **(some-other-link.md)** instead of **(projects/some-other-link.md)** will not work inside obsidian.nvim and will produce the error **Failed to resolve file 'some-other-link.md'** when calling :ObsidianFollowLink.

Am I doing something wrong here or is relative path in vault linking not supported by this extension (specifically :ObsidianFollowLink or gf)?

Thanks in advance

**Comment by epwalsh** [comment]
Hey @Woodshape, this is supported in general so it looks like you may have found a bug. Is your obsidian.nvim up to date with `main`?

**Comment by Woodshape** [comment]
Hey @epwalsh, thanks for the quick answer.

I am using the current master branch version 2.5.0. 
The problem / bug still persists with the 2.5.1. version.

The same thing happens with :ObsidianBacklinks not working, which makes sense in this case.

Btw. I am using the default config in my lazy.nvim setup.

**Reply by epwalsh** [reply]
If your notes go in a `projects` subfolder, do you have `notes_subdir` set to `"projects"` in your config?

**Reply by epwalsh** [reply]
Ah, I did find a bug though that should fix `:ObsidianFollowLink` for you regardless of what your `notes_subdir` is set to. 0f7517b

I'm still looking into the backlinks issue.

**Reply by epwalsh** [reply, marked as answer]
Backlinks issue now fixed as well: 827668e

**Reply by Woodshape** [reply]
Awesome!
Thanks for the great and quick work. 
I really appreciate that and love the work you (and the other contributors) are doing.

**Reply by morgaan** [reply]
Hi,

First of all thank you for this awesome plugin!
I have the same use case than @Woodshape, and would like my notes to be able to resolve backlinks even outside Obsidian and Neovim. I'm using currently version `3.6.1` and I'm using `markdown` links rather than `wiki` one.
I have a vault, with 5 folders (`0 Inbox`, `1 Projects`, `2 Areas`...) following the Second Brain PARA method. In the projects folder I have one level of subfolder for each project. I have a main note per project that aim at linking at other notes, both notes that lives beside the main note, or leaving somewhere else in the vault. What I'm aiming for is the ability from the `cmp` suggestions to have the options of a link with the relative path. At the moment, and this could be a misconfiguration from my end, all the suggestions contains an absolute link (e.g `1 Projects/My exciting new project/exciting-new-project-inspirations.md` (I would expect `exciting-new-project-inspirations.md` at that point). The worth part is that the Obsidian client is not able to link to this note when I click on its link from the main note. Moreover it attempts to create a new note in same folder with name `1`. If if manually patch the path to be relative after using the completion, both `gf` in the plugin and clicking the link in the client work. Moreover, the linkage also work in Gitlab, oh and the backlinks functionality then works as well.

Below is the configuration options I've set if this is of any help.
```lua
  opts = {
    workspaces = {
      {
        name = "Second Brain",
        path = "~/second-brain",
      }
    },

	-- Completion of wiki links, local markdown links, and tags using nvim-cmp.
	completion = {
		-- Set to false to disable completion.
		nvim_cmp = true,
		-- Trigger completion at 2 chars.
		min_chars = 2,
	},

	preferred_link_style = "markdown",

	picker = {
		name = "telescope.nvim",
		mappings = {
			-- Create a new note from your query.
			new = "<C-x>",
			-- Insert a link to the selected note.
			insert_link = "<C-l>",
		}
	},

	note_frontmatter_func = function(note)
		-- Add the title of the note as an alias.
		if note.title then
			note:add_alias(note.title)
		end

		local out = { aliases = note.aliases, tags = note.tags }

		-- `note.metadata` contains any manually added fields in the frontmatter.
		-- So here we just make sure those fields are kept in the frontmatter.
		if note.metadata ~= nil and not vim.tbl_isempty(note.metadata) then
			for k, v in pairs(note.metadata) do
				out[k] = v
			end
		end

		return out
	end,

	new_notes_location = "current_dir",

	note_id_func = function(title)
		-- Create note IDs in a Zettelkasten format with a timestamp and a suffix.
		-- In this case a note with the title 'My new note' will be given an ID that looks
		-- like '1657296016-my-new-note', and therefore the file name '1657296016-my-new-note.md'
		local suffix = ""
		if title ~= nil then
			-- If title is given, transform it into valid file name.
			suffix = title:gsub(" ", "-"):gsub("[^A-Za-z0-9-]", ""):lower()
		else
			-- If title is nil, just add 4 random uppercase letters to the suffix.
			for _ = 1, 4 do
				suffix = suffix .. string.char(math.random(65, 90))
			end
		end
		-- return tostring(os.time()) .. "-" .. suffix
		return suffix
	end,

	templates = {
		subdir = "4 Archives/zzz_notes-templates/nvim",
		date_format = "%Y-%m-%d",
		time_format = "%H:%M",
		tags = "",
		substitutions = {
			year = function()
				return os.date("%Y", os.time())
			end
		}
	},

	markdown_link_func = function(opts)
		return require("obsidian.util").markdown_link(opts)
	end,

	follow_url_func = function(url)
		-- Open the URL in the default web browser.
		if (vim.loop.os_uname().sysname == "Linux") then
			vim.fn.jobstart({"xdg-open", url}) -- Linux
		else
			vim.fn.jobstart({"open", url})  -- MacOS
		end
	end,
  }
```

Am I doing anything wrong?

Thank you in advance

**Reply by epwalsh** [reply]
Hey @morgaan...

> What I'm aiming for is the ability from the cmp suggestions to have the options of a link with the relative path.

Do the `cmp` suggestions you want, you just need to customize your `markdown_link_func`. See `obsidian.util.markdown_link()` for a starting point.

As for the issue with following those links, I just figured out what that issue was and pushed a fix: 3393e3b

**Reply by morgaan** [reply]
Thank you @epwalsh, unfortunately I'm not a Lua developer and although I looked at the code of `obsidian.util.markdown_link()` I'm not sure how to construct the `path` to be relative to the note currently loaded in the buffer.
Searching the source code for `relative_path` I see some occurences using a `client` instance that I'm not sure we have at disposal in the scope of that `markdown_link_func` function (override).
Maybe something more obvious for you :shrug:.
Moreover I feel this could be something that you could add to your recipes and that would most certainly benefit the community. Trust me I wish I would be enough equipped to benefit myself here, but I'm not :cry:.

Looking forward to your answer, and thank you again for all the hard work.

**Reply by epwalsh** [reply]
If you just want the filename for the path, then this would work:

```lua
markdown_link = function(opts)
  local fname = vim.fs.basename(opts.path)
  return string.format("[%s](%s)", opts.label, fname)
end,
```

**Reply by morgaan** [reply]
Thank you @epwalsh, it's a start but I would also like it to work if the file is not right besides the one from the current buffer.

I have just spent hours trying to come up with something, and this is the "closest" I came to. Unfortunately, it is still not working because of the `vim.fn.expand` being, from my understanding, an async operation.
I tried with `vim.schedule` and as you can see `vim.schedule_wrap` with no luck, also my understanding of it is really vague.
I'm stuck, what is very frustrating after all the effort and going outside my comfort zone.
I may be close and just missing the hints a lua and/or neovim plugin developer would have.

Worth noting, I've also updated my nvim from 0.9.2 to 0.10.0, in case this would have be an issue.

Hopefully it is an easy one for you, if you do not mind giving it a look.

```lua
markdown_link_func = vim.schedule_wrap(function(opts)
	local vault = '~/second-brain'
	local vaultAbsolutePath = vim.fn.fnamemodify(vault, ':p')
	local suggestionAbsolutePath = string.format("%s%s", vaultAbsolutePath, opts.path)

	if (vim.fn.filereadable(suggestionAbsolutePath) == 1) then
		local currentNoteFolderAbsolutePath = vim.fn.expand('%:p:h')
		local Path = require("plenary.path")
		local relpath = Path:new(suggestionAbsolutePath):make_relative(currentNoteFolderAbsolutePath)
		local path = require("obsidian.util").urlencode(relpath, { keep_path_sep = true })
		return string.format("[%s](%s)", opts.label, path)
	end

	local fname = vim.fs.basename(opts.path)
	return string.format("[%s](%s)", opts.label, fname)
end),
```

With this I do get the following stacktrace:

```
Error detected while processing TextChangedI Autocommands for "*":
Error executing lua callback: ...l/share/nvim/lazy/obsidian.nvim/lua/cmp_obsidian_new.lua:43: attempt to concatenate local 'new_text' (a nil value)
stack traceback:
	...l/share/nvim/lazy/obsidian.nvim/lua/cmp_obsidian_new.lua:43: in function 'complete'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/source.lua:326: in function 'complete'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:299: in function 'complete'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:169: in function 'callback'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:229: in function 'autoindent'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:161: in function 'on_change'
	...morgan/.local/share/nvim/lazy/nvim-cmp/lua/cmp/init.lua:338: in function 'callback'
	...local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/autocmd.lua:49: in function 'emit'
	...local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/autocmd.lua:23: in function <...local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/autocmd.lua:22>
Error executing luv callback:
...local/share/nvim/lazy/obsidian.nvim/lua/cmp_obsidian.lua:92: table index is nil
stack traceback:
	...local/share/nvim/lazy/obsidian.nvim/lua/cmp_obsidian.lua:92: in function 'callback'
	...al/share/nvim/lazy/obsidian.nvim/lua/obsidian/client.lua:477: in function 'callback'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:25: in function 'callback_or_next'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function 'saved_callback'
	...are/nvim/lazy/plenary.nvim/lua/plenary/async/control.lua:126: in function 'tx'
	...cal/share/nvim/lazy/obsidian.nvim/lua/obsidian/async.lua:62: in function 'callback'
	...cal/share/nvim/lazy/obsidian.nvim/lua/obsidian/async.lua:214: in function 'callback'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:25: in function 'callback_or_next'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function <...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:44>
```

Maybe there is another way to get `currentNoteFolderAbsolutePath` without the async operation :shrug:. Not sure I would not hit another wall after that :stuck_out_tongue: 
 
Thank you in advance

**Reply by epwalsh** [reply]
> because of the vim.fn.expand being, from my understanding, an async operation.

I think this is actually the opposite. `vim.fn.*` functions cannot be called from an async context, but the `markdown_link_func` is potentially called in an async context.

Anyway, try this:

```lua
markdown_link_func = function(opts)
  local util = require "obsidian.util"
  local Path = require "obsidian.path"
  local client = require("obsidian").get_client()

  -- This is now an absolute path to the file.
  local path = client.dir / opts.path
  -- This is an absolute path to the current buffer's parent directory.
  local buf_dir = Path.buf_dir()

  ---@type string
  local rel_path
  if buf_dir:is_parent_of(path) then
    -- If `path` is a child of `buf_dir`, return the path relative to `buf_dir`.
    rel_path = tostring(path:relative_to(buf_dir))
  else
    rel_path = path.name
  end

  local encoded_path = util.urlencode(rel_path, { keep_path_sep = true })
  return string.format("[%s](%s)", opts.label, encoded_path)
end,
```

**Reply by morgaan** [reply]
Hi @epwalsh this looks much clean than what I came up with. Also I did not catch that your plugin had all these neat utilities to get the client and some path tooling :clap.

Unfortunately we seem to get down to the same-ish issue.
Now I'm getting the following stacktrace:

```
Error executing luv callback:
...ocal/share/nvim/lazy/obsidian.nvim/lua/obsidian/path.lua:214: E5560: nvim_buf_get_name must not be called in a lua loop callback
stack traceback:
	[C]: in function 'nvim_buf_get_name'
	...ocal/share/nvim/lazy/obsidian.nvim/lua/obsidian/path.lua:214: in function 'buffer'
	...ocal/share/nvim/lazy/obsidian.nvim/lua/obsidian/path.lua:223: in function 'buf_dir'
	/home/morgan/.config/nvim/lua/morgan/plugins/obsidian.lua:104: in function 'format_link'
	...local/share/nvim/lazy/obsidian.nvim/lua/cmp_obsidian.lua:59: in function 'callback'
	...al/share/nvim/lazy/obsidian.nvim/lua/obsidian/client.lua:477: in function 'callback'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:25: in function 'callback_or_next'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function 'saved_callback'
	...are/nvim/lazy/plenary.nvim/lua/plenary/async/control.lua:126: in function 'tx'
	...cal/share/nvim/lazy/obsidian.nvim/lua/obsidian/async.lua:62: in function 'callback'
	...cal/share/nvim/lazy/obsidian.nvim/lua/obsidian/async.lua:214: in function 'callback'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:25: in function 'callback_or_next'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function <...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:44>
```

The line originating the error seem to be:

```lua
local buf_dir = Path.buf_dir()
```
which I believe may be using `nvim_buf_get_name` under the hood. The bottleneck seem to be getting information about current buffer in a "in a lua loop callback".
Not sure if there is a workaround at that point, or if the `opts` argument would contain, not only information about the suggested path, but also the current buffer that, may enable us to achieve that :shrug:.

Any idea?

Thank you in any case so much for your support so far. I sure the solution is around the corner :crossed_fingers: 

Looking forward to hearing from you.

Cheers

**Reply by epwalsh** [reply]
Ahhh darn. Actually there is another way to get the buf directory! The client stores this as a field, so replace `local buf_dir = Path.buf_dir()` with `local buf_dir = client.buf_dir`.

**Reply by morgaan** [reply]
Nice! Thank you @epwalsh, that is obviously super convenient. I think I brought it to the closest we can we tools at hand ATM.

```lua
	markdown_link_func = function(opts)
		local util = require "obsidian.util"
		local Path = require "obsidian.path"
		local client = require("obsidian").get_client()

		-- This is now an absolute path to the file.
		local path = Path:new(client.dir / opts.path)
		-- This is an absolute path to the current buffer's parent directory.
		local buf_dir = client.buf_dir

		---@type string
		local rel_path
		if buf_dir:is_parent_of(path) then
			-- If `path` is a child of `buf_dir`, return the path relative to `buf_dir`.
			rel_path = tostring(path:relative_to(buf_dir))
		else
			rel_path = path.name
		end

		local encoded_path = util.urlencode(rel_path, { keep_path_sep = true })
		return string.format("[%s](%s)", opts.label, rel_path)
	end,
``` 

Note: I've patch `local path = client.dir / opts.path` to  `local path = Path:new(client.dir / opts.path)`.

So far I understand that does the job only if the suggested has in its ancestors the parent directory of the current buffer. That does not do the job for suggestions located elsewhere unfortunately what may be my most used case apart from when notes live besides another. But But, looking at the `plenary.path` I found [this issue that suggest that this kind of relative path is not yet supported](https://github.com/nvim-lua/plenary.nvim/issues/411). If the behaviour of the `make_relative` method would be as intuitive as it suggests that would fix this once and for all I think.

As an interim solution, I believe I can:
1. Save and switch over to the Obsidian client with `:w | ObsidianOpen` (that I can map to a key mapping)
2. Move to where I want the link (may be the most tedious operation)
3. Add the link there (I've configured to use relative path in the Link preference)
4. Switch back to Vim and run `:e!` to load the changes in the buffer

Thank you so much for your tremendous contribution through this plugin as well as you support on my issue.
I'll watch the issue in `plenary` repository, who knows.

**Reply by epwalsh** [reply]
`local path = client.dir / opts.path` and `local path = Path:new(client.dir / opts.path)` should do the exact same thing. Also, we're using `obsidian.path`, not `plenary.path`. I implemented `obsidian.path` from scratch because `plenary.path` has several bugs and poorly defined behavior. `obsidian.path` tries to adhere to Python's pathlib API as close as possible.

> That does not do the job for suggestions located elsewhere unfortunately what may be my most used case apart from when notes live besides another.

Can you give me a concrete example of this situation? What's the absolute path to the current buffer, what's the absolute path to the new note, and what's the relative path to the new note that you'd like to see?

**Reply by morgaan** [reply]
Hi @epwalsh and thank you,

Let's take a concrete example:

```
second-brain (vault root)
  └── 0 Inbox
  └── 1 Projects
      └── Project A
        └── meeting minutes.md
        └── backlog.md
        └── Diagrams
           └── sequence diagram.md
       ...
  └── 3 References
      └── Website
        └── color palette.md
...
```

- If `meeting minutes.md` references `backlog.md` we would have `[Project A backlog](backlog.md)` :+1: 
- If `backlog.md` references `sequence diagram.md` we would have: `[Project A sequence diagram](Diagrams/sequence%20diagram.md)` :+1: 
- If `backlog.md` references `color palette.md` we would have `[Website's color palette](../../3%20References/Website/color%20palette.md)` :-1: 

The bullets with :+1: IIRC are covered by the latest `markdown_link_func`, the one with :-1: not.

Hopefully that makes sense. That is, at least, how the Obsidian app behaves when the preference `File and Links > New link Format (What links to insert when auto-generating internal links.)` is set to `Relative path to file`.

I hope that this example brings clarity :crossed_fingers:. Thank you so much for you tenacity :blush:.

Looking forward to you answer,

Cheers

**Reply by epwalsh** [reply]
Ah, okay, then I think this would work:

```lua
---@type string|?
local rel_path
if buf_dir:is_parent_of(path) then
  rel_path = tostring(path:relative_to(buf_dir))
else
  local parents = buf_dir:parents()
  for i, parent in ipairs(parents) do
    if parent:is_parent_of(path) then
      rel_path = string.rep("../", i) .. tostring(path:relative_to(parent))
      break
    end
  end
end

assert(rel_path)
```

**Reply by morgaan** [reply]
Hi @epwalsh you are a genious :clap: thank you so so much. We are almost there I think. Only one issue left is that weirdly the `util.urlencode()` function seems to behave the same no matter if `keep_path_sep` is `true` or `false`, and encoding the `/` chars as well, making the resulting path not working (e.g. `[quick capture](..%2F..%2F0%20Inbox%2F_quick-capture.md)`, when it should have been `[quick capture](../../0%20Inbox/_quick-capture.md)`)

Here is my very last version of the `markdown_link_func`, I may be missing something or using something wrongly:

```lua
	markdown_link_func = function(opts)
		local util = require "obsidian.util"
		local Path = require "obsidian.path"
		local client = require("obsidian").get_client()

		-- This is now an absolute path to the file.
		local path = client.dir / opts.path
		-- This is an absolute path to the current buffer's parent directory.
		local buf_dir = client.buf_dir

		local rel_path
		if buf_dir:is_parent_of(path) then
			rel_path = tostring(path:relative_to(buf_dir))
		else
			local parents = buf_dir:parents()
			for i, parent in ipairs(parents) do
				if parent:is_parent_of(path) then
					rel_path = string.rep("../", i) .. tostring(path:relative_to(parent))
					break
				end
			end
		end

		local encoded_path = util.urlencode(rel_path, { keep_path_sep = true })
		return string.format("[%s](%s)", opts.label, encoded_path)
	end,
```
I'm sure it's a tiny thing somewhere.

Looking forward to what I guess would be the last fix.
Moreover, I think this would make a great configuration recipe for ones looking to mimic the previously mentioned Obsidian app's `File and Links > New link Format (What links to insert when auto-generating internal links.) > Relative path` preference.

Cheers

**Reply by morgaan** [reply]
Oh hold on! I've just updated to very latest release (3.7.1) and it seems that the issue with `keep_path_sep` is gone :star_struck:.

This totally made my day! Thank you so so much @epwalsh. 5-stars support :blush:

And for anyone interested passing by here; this is my very latest version, including now `header` and `anchor` support:

```lua
	markdown_link_func = function(opts)
		local util = require "obsidian.util"
		local client = require("obsidian").get_client()
		local anchor = ""
		local header = ""

		if opts.anchor then
			anchor = opts.anchor.anchor
			header = util.format_anchor_label(opts.anchor)
		elseif opts.block then
			anchor = "#" .. opts.block.id
			header = "#" .. opts.block.id
		end

		-- This is now an absolute path to the file.
		local path = client.dir / opts.path
		-- This is an absolute path to the current buffer's parent directory.
		local buf_dir = client.buf_dir

		local rel_path
		if buf_dir:is_parent_of(path) then
			rel_path = tostring(path:relative_to(buf_dir))
		else
			local parents = buf_dir:parents()
			for i, parent in ipairs(parents) do
				if parent:is_parent_of(path) then
					rel_path = string.rep("../", i) .. tostring(path:relative_to(parent))
					break
				end
			end
		end

		local encoded_path = util.urlencode(rel_path, { keep_path_sep = true })
		return string.format("[%s%s](%s%s)", opts.label, header, encoded_path, anchor)
	end,
```

Thank you again so much Pete!

Cheers
